### PR TITLE
Add company creation functionality

### PR DESCRIPTION
**Pull request by RuslanKhabibullinFs**
https://trello.com/c/O2hWJfXj

### CODE REVIEW

File: app/forms/company_registration_form.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  validates :email, presence: true, format: { with: Devise.email_regexp }
+  validates :first_name, :last_name, presence: true
+  validates :company_subdomain, presence: true,
+                                uniqueness: { model: Company, attribute: "subdomain" },
+                                format: { with: COMPANY_SUBDOMAIN_REGEXP },
+                                exclusion: { in: Subdomains::Base::DEFAULT_SUBDOMAINS }
+  validates :company_name, presence: true,
+                           uniqueness: { case_sensitive: false, model: Company, attribute: "name" }
+  validates :password, :password_confirmation, presence: true, length: { in: Devise.password_length }
+  validate :password_equality_check
```
Comment (ArthurZaharov) [inline app/forms/company_registration_form.rb]: Can you replace this with `validates :password, confirmation: true`?

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  validates :first_name, :last_name, presence: true
+
+  def self.find_for_authentication(warden_conditions)
+    company = Company.find_by(subdomain: warden_conditions[:subdomain])
+    find_by(email: warden_conditions[:email], company: company)
+  end
+
+  private
+
+  def password_required?
```
Comment (ArthurZaharov) [inline app/models/user.rb]: Why you need this check? If you need it to update user info then maybe you can use Devise build in functionality - https://github.com/plataformatec/devise/wiki/How-To:-Allow-users-to-edit-their-account-without-providing-a-password

File: app/validators/uniqueness_validator.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+class UniquenessValidator < ActiveRecord::Validations::UniquenessValidator
```
Comment (ArthurZaharov) [inline app/validators/uniqueness_validator.rb]: It isn't good idea to rewrite rails build in validator definitely.

File: app/views/company_registrations/new.html.slim [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    h2 Company Registration
+
+    = simple_form_for(company_registration_form,
+        url: company_registration_path(company_registration_form),
+        html: { class: "company-registration-form" },
+        defaults: { label: false }) do |f|
+
+      = f.error_notification
+
+      .form-inputs
```
Comment (ArthurZaharov) [inline app/views/company_registrations/new.html.slim]: Can you add this classes in SimpleForm wrapper?

File: app/views/layouts/_flash_messages.html.slim [hunk 5]
```diff
@@ -0,0 +1,4 @@
+.info-block
+  - flash.each do |key, value|
+    div[class = "alert alert-#{key == "notice" ? "info" : "warning"}"]
```
Comment (ArthurZaharov) [inline app/views/layouts/_flash_messages.html.slim]: If you need this key transformation for Bootstrap styles then you can try to do it from css side like suggested here - https://github.com/plataformatec/devise/issues/2051#issuecomment-254652729

File: lib/subdomains/base.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+module Subdomains
```
Comment (ArthurZaharov) [inline lib/subdomains/base.rb]: Why you paste it in `lib` folder? In `lib` folder usually stored vendor code libraries. All code that you write should be in `app` folder

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -1,12 +1,18 @@
...
 
   self.responder = ApplicationResponder
 
   protect_from_forgery with: :exception
 
-  responders :flash
   respond_to :html
+  responders :flash
+
+  def after_sign_in_path_for(_record)
```
Comment (ArthurZaharov) [inline app/controllers/application_controller.rb]: Do you really need this method? Does `Devise` standard behaviour enough?

File: app/controllers/company_registrations_controller.rb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  delegate :owner, to: :company_registration_form
+  delegate :company, to: :owner, prefix: true
+
+  def new
+  end
+
+  def create
+    if company_registration_form.save
+      sign_in(owner)
+      return redirect_to(posts_url(subdomain: owner_company.subdomain), notice: t(".success"))
```
Comment (ArthurZaharov) [inline app/controllers/company_registrations_controller.rb]: You shouldn't use return in controller actions.

**Review by ArthurZaharov — CHANGES_REQUESTED** [review]

File: app/models/user.rb (hunk 2 above)
Comment (RuslanKhabibullinFs) [reply app/models/user.rb]: After some investigation i found that this method is needed) Without it is impossible to implement the functional: the password is required during registration process. And is not needed during edit/update profile process. (I get this method from Devise validatable module - https://github.com/plataformatec/devise/blob/master/lib/devise/models/validatable.rb)

**Pull request merged**